src/wp-admin/includes/file.php

```diff
@@ -1946,6 +1946,75 @@ function copy_dir( $from, $to, $skip_list = array() ) {
 	return true;
 }
 
+/**
+ * Moves a directory from one location to another.
+ *
+ * Recursively invalidates OPcache on success.
+ *
+ * If the renaming failed, falls back to copy_dir().
+ *
+ * Assumes that WP_Filesystem() has already been called and setup.
+ *
+ * @since 6.2.0
+ *
+ * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+ *
+ * @param string $from      Source directory.
+ * @param string $to        Destination directory.
+ * @param bool   $overwrite Optional. Whether to overwrite the destination directory if it exists.
+ *                          Default false.
+ * @return true|WP_Error True on success, WP_Error on failure.
+ */
+function move_dir( $from, $to, $overwrite = false ) {
+	global $wp_filesystem;
+
+	if ( trailingslashit( strtolower( $from ) ) === trailingslashit( strtolower( $to ) ) ) {
+		return new WP_Error(
+			'source_destination_same_move_dir',
+			__( 'The source and destination are the same.' )
+		);
+	}
+
+	if ( ! $overwrite && $wp_filesystem->exists( $to ) ) {
+		return new WP_Error( 'destination_already_exists_move_dir', __( 'The destination folder already exists.' ), $to );
+	}
+
+	if ( $wp_filesystem->move( $from, $to, $overwrite ) ) {
+		/*
+		 * When using an environment with shared folders,
+		 * there is a delay in updating the filesystem's cache.
+		 *
+		 * This is a known issue in environments with a VirtualBox provider.
+		 *
+		 * A 200ms delay gives time for the filesystem to update its cache,
+		 * prevents "Operation not permitted", and "No such file or directory" warnings.
+		 *
+		 * This delay is used in other projects, including Composer.
+		 * @link https://github.com/composer/composer/blob/2.5.1/src/Composer/Util/Platform.php#L228-L233
+		 */
+		usleep( 200000 );
+		wp_opcache_invalidate_directory( $to );
```

Comment:
Do we really need to have the call to `wp_opcache_invalidate_directory()` here?

FWIW, it seems inconsistent to me that this check is added here, but not used after any other file system modification. What makes this function different from other file system modifications?

Since this isn't handled anywhere else in core, shouldn't this rather be plugin territory or alternatively handled by core consistently?

Reply:
`copy_dir` runs `wp_opcache_invalidate()` on every file. We do the same using this function. It also helps with VirtualBox compatibility. 

Loads of discussion on https://github.com/WordPress/wordpress-develop/pull/3798

Reply:
> Do we really need to..

Yes. Many random/unpredictable errors can happen when upgrading .php files without invalidating the opcode cache, especially on busy sites. See https://github.com/WordPress/wordpress-develop/pull/3798#issuecomment-1377946094.